xapian-letor/scorer/err_score.cc

```diff
@@ -50,42 +50,44 @@ ERRScore::score(const std::vector<FeatureVector> & fvv) const
 {
     LOGCALL(API, double, "ERRScore::score", fvv);
 
-    // hard code for the a five-point scale, the 16 means 2^(5-1)
-    int MAX_LABEL = 16;
+    int MAX_PROB = 16;
 
-    std::vector<double> labels;
+    int length = fvv.size();
+
+    double vector_intermediate[length];
     double max_label = fvv[0].get_label();
 
-    for (size_t i = 0; i < fvv.size(); ++i) {
+    // store the labels set the by the user in vector_intermediate.
+    for (int i = 0; i < length; ++i) {
 	double label = fvv[i].get_label();
-	labels.push_back(fvv[i].get_label());
+	vector_intermediate[i] = label;
 	max_label = max(max_label, label);
     }
 
-    for (size_t i = 0; i < fvv.size(); ++i) {
-	labels[i] = round((labels[i] * 4) / max_label);
+    // normalize the labels to an integer from 0 to 4 because we are using
+    // a grade point system where each document is graded from 0 to 4.
+    for (int i = 0; i < length; ++i) {
+	vector_intermediate[i] = round((vector_intermediate[i] * 4) /
+				       max_label);
     }
 
-    int length = labels.size();
-
-    // compute the satisfaction probability for lable of each doc in the ranking
+    // compute the satisfaction probability for label of each doc in the ranking.
     for (int i = 0; i < length; ++i) {
-	labels[i] = (pow(2, labels[i]) - 1) / MAX_LABEL;
+	vector_intermediate[i] = (pow(2, vector_intermediate[i]) - 1) /
+				  MAX_PROB;
```

Comment:
Since we've collapsed the label space to `{0, 1, 2, 3, 4}`, we should probably pre-calculate the possible values of this expression, so we aren't doing O(n) `pow()` calls. (In fact, we can then do no `pow()` calls at all, using an accumulator in the pre-calculation loop.)

Reply:
FWIW, `pow(2, x)` will very likely get optimised to `x * x` and not call `pow()` at all (IIRC, both GCC and clang do this).

Reply:
Oh, that's handy. Although an accumulator should still be faster overall, it won't be as clear.